lib/active_record/connection_adapters/sqlserver/database_statements.rb

```diff
@@ -140,6 +140,37 @@ def default_insert_value(column)
         private :default_insert_value
 
         def build_insert_sql(insert) # :nodoc:
+          if insert.skip_duplicates?
+            # Do we have a unique_by index? Use index columns
+            conflict_columns = if (unique_by = insert.send(:insert_all).unique_by)
+              [unique_by.columns]
+            else
+              # Compare against every unique constraint (primary key included).
+              # Discard constraints that are not fully included on insert.keys. Prevents invalid queries.
+              # Example: ignore unique index for columns ["name"] if insert keys is ["description"]
+              insert_all = insert.send(:insert_all)
+
+              (insert_all.send(:unique_indexes).map(&:columns) + [insert_all.primary_keys]).select do |columns|
+                columns.to_set.subset?(insert.keys)
+              end
+            end
+            includes_primary_key = (insert.send(:insert_all).primary_keys.to_set & insert.keys).present?
+
+            sql = +""
+            sql << "SET IDENTITY_INSERT #{insert.model.quoted_table_name} ON;" if includes_primary_key
+            sql << "MERGE INTO #{insert.model.quoted_table_name} WITH (UPDLOCK, HOLDLOCK) AS target"
```

Comment:
I believe HOLDLOCK is equivalent to SERIALIZABLE. I'd prefer the SQL standard term. Unless HOLDLOCK is usual in the SQL Server community?

Reply:
I will check that. Thanks!